exporter/awsprometheusremotewriteexporter/factory.go

```diff
@@ -0,0 +1,127 @@
+func NewFactory() component.ExporterFactory {
+	return &AwsFactory{ExporterFactory: prw.NewFactory()}
+	// return exporterhelper.NewFactory(
+	// 	typeStr,
+	// 	createDefaultConfig,
+	// 	exporterhelper.WithMetrics(createMetricsExporter))
+}
+
+func (af *AwsFactory) Type() configmodels.Type {
+	return typeStr
+}
+
+func (af *AwsFactory) CreateMetricsExporter(ctx context.Context, params component.ExporterCreateParams,
+	cfg configmodels.Exporter) (component.MetricsExporter, error) {
+	prwCfg := cfg.(*Config)
+
+	prwCfg.HTTPClientSettings.CustomRoundTripper = func(next http.RoundTripper) (http.RoundTripper, error) {
+		if !validateAuthConfig(prwCfg.AuthSettings) {
+			return nil, errors.New("invalid authentication configuration")
+		}
+
+		return newSigningRoundTripper(prwCfg.AuthSettings, next)
+	}
+
+	// return af.ExporterFactory.CreateMetricsExporter(ctx, params, prwCfg)
+	client, cerr := prwCfg.HTTPClientSettings.ToClient()
+	if cerr != nil {
+		return nil, cerr
+	}
+
+	// initialize an upstream exporter and pass it an http.Client with interceptor
+	prwe, err := prw.NewPrwExporter(prwCfg.Namespace, prwCfg.HTTPClientSettings.Endpoint, client, prwCfg.ExternalLabels)
+	if err != nil {
+		return nil, err
+	}
+
+	// use upstream helper package to return an exporter that implements the required interface, and has timeout,
+	// queueing and retry feature enabled
+	prwexp, err := exporterhelper.NewMetricsExporter(
+		cfg,
+		params.Logger,
+		prwe.PushMetrics,
+		exporterhelper.WithTimeout(prwCfg.TimeoutSettings),
+		exporterhelper.WithQueue(prwCfg.QueueSettings),
+		exporterhelper.WithRetry(prwCfg.RetrySettings),
+		exporterhelper.WithShutdown(prwe.Shutdown),
+	)
+
+	return prwexp, err
+}
+
+func (af *AwsFactory) CreateDefaultConfig() configmodels.Exporter {
+	qs := exporterhelper.CreateDefaultQueueSettings()
+	qs.Enabled = false
+
+	ts := exporterhelper.CreateDefaultRetrySettings()
+	ts.Enabled = false
+
+	cfg := &Config{
+		Config: prw.Config{
+			ExporterSettings: configmodels.ExporterSettings{
+				TypeVal: typeStr,
+				NameVal: typeStr,
+			},
+			Namespace:       "",
+			ExternalLabels:  map[string]string{},
+			TimeoutSettings: exporterhelper.CreateDefaultTimeoutSettings(),
+			RetrySettings:   ts,
+			QueueSettings:   qs,
+			HTTPClientSettings: confighttp.HTTPClientSettings{
+				Endpoint: "http://some.url:9411/api/prom/push",
+				// We almost read 0 bytes, so no need to tune ReadBufferSize.
+				ReadBufferSize:  0,
+				WriteBufferSize: 512 * 1024,
+				Timeout:         exporterhelper.CreateDefaultTimeoutSettings().Timeout,
+				Headers:         map[string]string{},
+			},
+		},
+		AuthSettings: AuthSettings{
+			Region:  "",
+			Service: "",
+		},
+	}
+
+	return cfg
+}
+
+func validateAuthConfig(params AuthSettings) bool {
+	return !(params.Region != "" && params.Service == "" || params.Region == "" && params.Service != "")
```

Comment:
I'm tired... but I swear I saw nearly the same function in a previous file

Reply:
it is actually a slightly different function. This function checks if both fields are either filled or empty then it is valid (both empty meaning do not apply auth, both filled meaning apply auth) and any other combination is invalid. The other function then checks between the 2 valid configurations if both fields are specified, perform the auth, otherwise skip auth

Reply:
Nvm, was tired, I see that its different now